### PR TITLE
Add sparsity tutorial

### PR DESCRIPTION
**Pull request by george-qi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #67

Differential Revision: [D37257029](https://our.internmc.facebook.com/intern/diff/D37257029)

### CODE REVIEW

**Comment by george-qi** [comment]
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37257029).

File: docs/source/notebooks/sparse.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+In according with Principle #1, a MaskedTensor is created by passing in two sparse tensors, which can be initialized with any of the constructors, e.g. `torch.sparse_coo_tensor`.
+
+```{code-cell} ipython3
+i = [[0, 1, 1],
+     [2, 0, 2]]
+v =  [3, 4, 5]
+m = torch.tensor([True, False, True])
+
+values = torch.sparse_coo_tensor(i, v, (2, 3))
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:67]: I think it's easier to use `to_sparse()` as a constructor.

**Review by cpuhrsch — COMMENTED** [review]
Generally this tutorial looks great! I'm mostly adding comments on phrasing and some clarifications on my end, so please view this more as a discussion.

File: docs/source/notebooks/sparse.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
++++
+
+## Introduction
+
++++
+
+[Sparsity in PyTorch](https://pytorch.org/docs/stable/sparse.html) is a quickly growing area that has found a lot of support and demand due to its efficiency in both memory and compute. This tutorial is meant to be used in conjunction with the the PyTorch link above, as the sparse tensors are ultimately the building blocks for MaskedTensors (just as regular `torch.Tensor`s are as well).
+
+Sparse storage formats are particularly powerful in scenarios where the majority of elements are equal to zero. There are a number of different [sparse storage formats](https://en.wikipedia.org/wiki/Sparse_matrix) that can be leveraged with various tradeoffs and degrees of adoption.
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:28]: Not always true. Block sparse memory formats also work well when there's a lower degree of sparsity (number of zeros).

File: docs/source/notebooks/sparse.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+Noting PyTorch's terminology of "specified" and "unspecified" elements (e.g. the elements that are actually stored vs. not), the parallel to MaskedTensor's usage is clear. However, by allowing a mask as well, MaskedTensors are even more generalizable, as we'l show through the tutorial - e.g. when the mask all `True`, most operations will result in the same result, but when the mask indicates unspecified values, then values in the sparse tensor will be masked out.
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:30]: I think you can make a much stronger statement here. MaskedTensor is saving PyTorch core from the pain that is "specified" and "unspecified" elements. We've had masked semantics within core without ever saying so and without being consistent. We'll eventually remove this section from the core documentation entirely and instead point users to this project when they are looking for the ability to express "unspecified" elements.

File: docs/source/notebooks/sparse.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+<div class="alert alert-info">
+
+**Note:** Currently, only the COO sparse storage format is supported in MaskedTensor ([CSR is being developed](https://github.com/pytorch/maskedtensor/pull/65)). If you have another format that you would like supported, please file an issue!
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:34]: It seems worthwhile for CSR to get in first before landing this tutorial, since it's a very commonly requested format and I fear users might stop reading if they don't see it supported.

File: docs/source/notebooks/sparse.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+## Principles
+
++++
+
+1. `input` and `mask` must have the same storage format, whether that's `torch.strided`, `torch.sparse_coo`, or `torch.sparse_csr`.
+
+2. `input` and `mask` must have the same size, indicated by `t.size()`
+
+3. `input` and `mask` - only for sparse formats - can have a different number of elements (`tensor.nnz()`) **at creation**, the indices of `mask` must then be a subset of the indices from `input`. In this case, `input` will assume the shape of mask using the function `input.sparse_mask(mask)`; in other words, any of the elements in `input` that are not `True` in `mask` will be thrown away
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:48]: I think explicitly mentioning this case is a bit confusing. I'd rather the user doesn't have to worry about this at all even if we have to extent the capabilities of `sparse_mask` first.

File: docs/source/notebooks/sparse.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+## Supported Operations
+
++++
+
+### Unary
+
++++
+
+All unary operations are supported; for a list of operations, please refer to [here](https://pytorch.org/maskedtensor/main/unary.html).
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:116]: Nit: The way I'd write this is
```
[All unary operations are supported](https://pytorch.org/maskedtensor/main/unary.html)
```
I generally like to avoid using the "here" style of providing links, but maybe that's actually a bad thing.

File: docs/source/notebooks/sparse.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+print("mt2:\n", mt2)
+print("torch.div(mt2, mt1):\n", torch.div(mt2, mt1))
+print("torch.mul(mt1, mt2):\n", torch.mul(mt1, mt2))
+```
+
+### Reductions
+
++++
+
+Unfortunately, only reductions across all dimensions are supported and not a particular dimension (e.g. `mt.sum()` is supported but not `mt.sum(dim=1)`). For a list of supported reductions, please refer [here](https://pytorch.org/maskedtensor/main/reductions.html).
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:154]: That shouldn't be the case anymore by using [_masked reductions](https://github.com/pytorch/pytorch/blob/master/torch/_masked/__init__.py#L826)

File: docs/source/notebooks/sparse.md (hunk 1 above)
Comment (george-qi) [reply docs/source/notebooks/sparse.md:78]: Changed!

File: docs/source/notebooks/sparse.md (hunk 2 above)
Comment (george-qi) [reply docs/source/notebooks/sparse.md]: Good point, added in this as a use case

File: docs/source/notebooks/sparse.md (hunk 4 above)
Comment (george-qi) [reply docs/source/notebooks/sparse.md]: CSR has been added with https://github.com/pytorch/maskedtensor/pull/65 :)

File: docs/source/notebooks/sparse.md (hunk 5 above)
Comment (george-qi) [reply docs/source/notebooks/sparse.md]: Removed this from "Principles" and moved it to an additional note below sparse COO tensors, since it's probably still interesting to know but indeed confusing as a core principle

File: docs/source/notebooks/sparse.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+values = torch.sparse_coo_tensor(i, v, (2, 3))
+mask = torch.sparse_coo_tensor(i, m, (2, 3))
+
+the following is simply shorter to write / easier to use
+"""
+
+values = torch.tensor([[0, 0, 3], [4, 0, 5]]).to_sparse()
+mask = torch.tensor([[False, False, True], [False, False, True]]).to_sparse()
+
+mt = masked_tensor(values, mask)
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:87]: Can't we also do `mt.to_sparse()`? The mask would be converted via `to_sparse` and the values filtered via `sparse_mask`? It seems like a more gentle introduction instead of constructing a MaskedTensor from the sparse components and dealing with differences in `nnz`.

File: docs/source/notebooks/sparse.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+
++++
+
+Similarly, MaskedTensor also supports the [CSR (Compressed Sparse Row)](https://pytorch.org/docs/stable/sparse.html#sparse-csr-tensor) sparse tensor format. Instead of storing the tuples of the indices like sparse COO tensors, sparse CSR tensors aim to decrease the memory requirements by storing compressed row indices. In particular, a CSR sparse tensor consists of three 1-D tensors:
+
+- `crow_indices`: array of compressed row indices with size `(size[0] + 1,)`. This array indicates which row a given entry in `values` lives in. The last element is the number of specified elements, while `crow_indices[i+1] - crow_indices[i]` indicates the number of specified elements in row `i`.
+- `col_indices`: array of size `(nnz,)`. Indicates the column indices for each value.
+- `values`: array of size `(nnz,)`. Contains the values of the CSR tensor.
+
+Of note, sparse CSR tensors are in a [beta](https://pytorch.org/docs/stable/index.html) state, so they are less developed in functionality than that of sparse COO tensors.
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:136]: nit: Technically both COO and CSR are in beta (for COO since 2016) and underdeveloped

File: docs/source/notebooks/sparse.md [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+```{code-cell} ipython3
+crow_indices = torch.tensor([0, 2, 4])
+col_indices = torch.tensor([0, 1, 0, 1])
+values = torch.tensor([1, 2, 3, 4])
+mask_values = torch.tensor([True, False, False, True])
+
+csr = torch.sparse_csr_tensor(crow_indices, col_indices, values, dtype=torch.double)
+mask = torch.sparse_csr_tensor(crow_indices, col_indices, mask_values, dtype=torch.bool)
+
+mt = masked_tensor(csr, mask)
```
Comment (cpuhrsch) [inline docs/source/notebooks/sparse.md:149]: Same thing as above, can we use `to_sparse_csr()`?

**Review by cpuhrsch — APPROVED** [review]

File: docs/source/notebooks/sparse.md (hunk 10 above)
Comment (george-qi) [reply docs/source/notebooks/sparse.md]: Done and moved this example to the Appendix!

File: docs/source/notebooks/sparse.md (hunk 8 above)
Comment (george-qi) [reply docs/source/notebooks/sparse.md]: Done and moved the more complex example to the bottom!

**Comment by george-qi** [comment]
@george-qi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37257029).
